File: solutions/sonnet-1/sonnet/java/brackets.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder output = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            output.append(isBalanced(line) ? "yes" : "no").append('\n');
        }
        System.out.print(output);
    }

    private static boolean isBalanced(String line) {
        Deque<Character> stack = new ArrayDeque<>();
        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            switch (c) {
                case '(': case '[': case '{':
                    stack.push(c);
                    break;
                case ')':
                    if (stack.isEmpty() || stack.pop() != '(') return false;
                    break;
                case ']':
                    if (stack.isEmpty() || stack.pop() != '[') return false;
                    break;
                case '}':
                    if (stack.isEmpty() || stack.pop() != '{') return false;
                    break;
                default:
                    break;
            }
        }
        return stack.isEmpty();
    }
}

File: solutions/sonnet-1/sonnet/javascript/brackets.js
const lines = require('fs').readFileSync(0, 'utf8').split('\n');

if (lines.length > 0 && lines[lines.length - 1] === '') {
  lines.pop();
}

const pairs = { ')': '(', ']': '[', '}': '{' };
const openers = new Set(['(', '[', '{']);

const results = lines.map((line) => {
  const stack = [];
  for (const ch of line) {
    if (openers.has(ch)) {
      stack.push(ch);
    } else if (pairs[ch]) {
      if (stack.pop() !== pairs[ch]) return 'no';
    }
  }
  return stack.length === 0 ? 'yes' : 'no';
});

console.log(results.join('\n'));
